File: strings/medium/3Longest Substring Without Repeating Characters
3LongestSubstringWithoutRepeatingCharacters/solution.py
def length_of_longest_substring(s):
    last = {}
    left = 0
    best = 0
    for i, ch in enumerate(s):
        if ch in last and last[ch] >= left:
            left = last[ch] + 1
        last[ch] = i
        best = max(best, i - left + 1)
    return best

File: strings/medium/3Longest Substring Without Repeating Characters
3LongestSubstringWithoutRepeatingCharacters/solution.java
class Solution {
  public int lengthOfLongestSubstring(String s) {
    int start = 0, end = 0, ans = 0;
    Set<Character> set = new HashSet<>();
    while (end < s.length()) {
      if (!set.contains(s.charAt(end))) {
        set.add(s.charAt(end++));
        ans = Math.max(ans, set.size());
      } else {
        set.remove(s.charAt(start++));
      }
    }
    return ans;
  }
}
